src/main/scala/translating/SemanticsLoader.scala

```diff
@@ -0,0 +1,521 @@
+  private var blockCount = 0
+
+  val opcodeSize = 4
+
+  def visitBlock(blockUUID: ByteString, blockCountIn: Int, blockAddress: Option[Int]): ArrayBuffer[Statement] = {
+    blockCount = blockCountIn
+    instructionCount = 0
+    val instructions = parserMap(Base64.getEncoder.encodeToString(blockUUID.toByteArray))
+
+    val statements: ArrayBuffer[Statement] = ArrayBuffer()
+
+    for (instruction <- instructions) {
+      constMap.clear
+      varMap.clear
+
+      for ((s, i) <- instruction.zipWithIndex) {
+        
+        val label = blockAddress.map {(a: Int) => 
+          val instructionAddress = a + (opcodeSize * instructionCount)
+          instructionAddress.toString + "$" + i
+        }
+        
+        val statement = visitStmt(s, label)
+        if (statement.isDefined) {
+          statements.append(statement.get)
+        }
+      }
+      instructionCount += 1
+    }
+    statements
+  }
+
+  private def visitStmt(ctx: StmtContext, label: Option[String] = None): Option[Statement] = {
+    ctx match {
+      case a: AssignContext => visitAssign(a, label)
+      case c: ConstDeclContext => visitConstDecl(c, label)
+      case v: VarDeclContext => visitVarDecl(v, label)
+      case v: VarDeclsNoInitContext =>
+        visitVarDeclsNoInit(v)
+        None
+      case a: AssertContext => visitAssert(a, label)
+      case t: TCallContext => visitTCall(t, label)
+      case i: IfContext => visitIf(i, label)
+      case t: ThrowContext => Some(visitThrow(t, label))
+    }
+  }
+
+  private def visitAssert(ctx: AssertContext, label: Option[String] = None): Option[Assert] = {
+    val expr = visitExpr(ctx.expr)
+    if (expr.isDefined) {
+      Some(Assert(expr.get, None, label))
+    } else {
+      None
+    }
+  }
+
+  private def visitThrow(ctx: ThrowContext, label: Option[String] = None): Assert = {
+    val message = ctx.ID().asScala.map(_.getText).mkString(" ,")
+    Assert(FalseLiteral, Some(message), label)
+  }
+
+  private def visitTCall(ctx: TCallContext, label: Option[String] = None): Option[Statement] = {
+    val function = ctx.ID.getText
+
+    val typeArgs = Option(ctx.tes) match {
+      case Some(e) => e.expr().asScala
+      case None => Nil
+    }
+    val args = Option(ctx.args) match {
+      case Some(e) => e.expr().asScala
+      case None => Nil
+    }
+
+    function match {
+      case "Mem.set.0" =>
+        checkArgs(function, 1, 4, typeArgs.size, args.size, ctx.getText)
+        val mem = Memory("mem", 64, 8) // yanked from BAP
+        val size = parseInt(typeArgs.head) * 8
+        val index = visitExpr(args.head)
+        val value = visitExpr(args(3))
+        val otherSize = parseInt(args(1)) * 8
+        val mysteryArg = parseInt(args(2))
+        if (size != otherSize) {
+          throw Exception(s"inconsistent size parameters in Mem.set.0: ${ctx.getText}")
+        }
+        if (mysteryArg != 0) {
+          Logger.debug(s"mystery 3rd arg of Mem.set.0 has value $mysteryArg: ${ctx.getText}")
```

Comment:
The 3rd arg is AccessType, maybe worth being less strict since we could see vector and atomic? It guess it hasn't shown up in practice but maybe worth commenting. 

```
enumeration AccType {AccType_NORMAL, AccType_VEC,        // Normal loads and stores
                     AccType_STREAM, AccType_VECSTREAM,  // Streaming loads and stores
                     AccType_ATOMIC, AccType_ATOMICRW,   // Atomic loads and stores
                     AccType_ORDERED, AccType_ORDEREDRW, // Load-Acquire and Store-Release
                     AccType_ORDEREDATOMIC,              // Load-Acquire and Store-Release with atomic access
                     AccType_ORDEREDATOMICRW,
                     AccType_LIMITEDORDERED,             // Load-LOAcquire and Store-LORelease
                     AccType_UNPRIV,                     // Load and store unprivileged
                     AccType_IFETCH,                     // Instruction fetch
                     AccType_PTW,                        // Page table walk
                     AccType_NONFAULT,                   // Non-faulting loads
                     AccType_CNOTFIRST,                  // Contiguous FF load, not first element
                     AccType_NV2REGISTER,                // MRS/MSR instruction used at EL1 and which is converted
                                                         // to a memory access that uses the EL2 translation regime
                     // Other operations
                     AccType_DC,                         // Data cache maintenance
                     AccType_DC_UNPRIV,                  // Data cache maintenance instruction used at EL0
                     AccType_IC,                         // Instruction cache maintenance
                     AccType_DCZVA,                      // DC ZVA instructions
                     AccType_AT};                        // Address translation
```

Reply:
That seems like a reason to throw an exception if we encounter another value, since we have not encountered any others yet and they may require a different translation. Where are the semantics of Mem.set and Mem.read defined? They aren't something that directly exists in the original ASL

Reply:
Its the implementation of the `Mem[]` array access operator https://github.com/UQ-PAC/aslp/blob/partial_eval/mra_tools/arch/arch.asl#L11795, and the actual name comes from the internal implementation of the interpreter https://github.com/UQ-PAC/aslp/blob/partial_eval/libASL/tcheck.ml#L1663

Reply:
Thanks for that. Looking at that I feel comfortable enough with the approach of just logging it for further investigation if we encounter a non-0 value.

Reply:
Yeah I agree that should be enough